explainer/effect-generic-bounds-and-functions.md

```diff
@@ -0,0 +1,544 @@
+if that's not enough to unambiguously figure out which variant to select, it
+will always be possible to explicitly tell the compiler which variant we
+expected using turbofish notation.
+
+```rust
+#[maybe(async)]
+fn meow() {
+    println!("meow");
+}
+
+async fn main() {
+    let fut = meow::<async>();   // `fn meow` is async
+    meow::<#[not(async)]>();     // `fn meow` is not async
+}
+```
+
+## Effect generic provided trait methods
+
+This RFC not just enables free functions to be generic over effects: it also
+enables provided trait methods to work with effect generics. Take for example
+our maybe-async trait `Read` again. By default it provides a number of methods,
+including `read_to_end`. With effect generic trait definitions, those provided
+functions can be generic over effects, meaning they can be made available to the
+effectful and base variants of the trait alike.
+
+```rust
+#[maybe(async)]
+pub trait Read {
+    // Using RFC 0000 required trait methods gained
+    // support for `#[maybe(effect)]` annotations.
+    #[maybe(async)]
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
+
+    // With this RFC, entire functions can be made
+    // generic over effects, meaning provided functions
+    // in traits now also work with `#[maybe(effect)]`.
+    #[maybe(async)]
+    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> Result<usize> { .. }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## TODO: Lowering
+
+Let's continue with our earlier example of the trait `Iterator` and the function
+`sum` which both have conditional support for the `async` effect via
+`#[maybe(async)]`. In its base form the trait `Iterator` looks like this:
+
+```rust
+/// The `Iterator` trait
+#[maybe(async)]
+pub trait Iterator {
+    type Item;
+    #[maybe(async)]
+    fn next(&mut self) -> Option<Self::Item>;
+}
+```
+
+Which using the desugaring proposed in RFC 0000, would desugar to a trait with a
+const-generic bool which determines whether it is async or not:
+
+```rust
+// Lowered trait definition
+pub trait Iterator<const IS_ASYNC: bool = false> {
+    type Item;
+    type Ret = Option<Self::Item>;
+    fn next(&mut self) -> Ret;
+}
+```
+
+We've seen how when maybe-async traits are implemented as sync they return their
+base type, and when implemented as async they wrap that up in an `impl Future` -
+see RFC 0000's section on "Lowering" for more details. Now for our function
+body, the base definition looks like this:
+
+```rust
+#[maybe(async)]
+pub fn sum<I>(iter: I) -> u32
+where
+    I: #[maybe(async)] Iterator<Item = u32>
+{
+    let mut total = 0;
+    while let Some(n) = iter.next().await {
+        total += n;
+    }
+    total
+}
+```
+
+TODO: ask oli for more details about the desugaring. Ref is: https://github.com/yoshuawuyts/maybe-async-channel/blob/2fc6fa012830525482d62a8facfae5e5a5e762fe/maybe-async-std/src/lib.rs#L47-L54
```

Comment:
@oli-obk this is the code I would like us to spell out the lowering for. Could you do that?